File: jsrc/legacy/SxxMachine/builtin/PRED_reverse_2.java
package SxxMachine.builtin;

import SxxMachine.ListTerm;
import SxxMachine.Operation;
import SxxMachine.Predicate;
import SxxMachine.Prolog;
import SxxMachine.SymbolTerm;
import SxxMachine.Term;
import SxxMachine.TermData;


public final class PRED_reverse_2 extends Predicate.P2 {
    private static final SymbolTerm s1 = SymbolTerm.intern("[]");

    public PRED_reverse_2(Term a1, Term a2, Operation cont) {
        this.LARG[0] = a1;
        this.LARG[1] = a2;
        this.cont = cont;
    }

    @Override
    public void toString(StringBuilder sb) {
        sb.append("reverse(");
        LARG[0].toString(sb); sb.append(", "); LARG[1].toString(sb);
        sb.append(")");
    }


    @Override
    public Operation exec(Prolog engine) {
        engine.setB0();
        Term a1 = LARG[0].dref();
        Term a2 = LARG[1].dref();

		if (a1 instanceof ListTerm || a1.isNil()) {
            Term r = Prolog.Nil;
            while (a1 instanceof ListTerm) {
                r = TermData.CONS(((ListTerm) a1).car(), r);
                a1 = ((ListTerm) a1).cdr();
            }
            if (a1.isNil() && a2.unify(r, engine.trail)){
            	return cont;
			}
        }
        return engine.fail();
    }
}

File: src/legacy/SxxMachine/builtin/PRED_reverse_2.cpp
using namespace std;

#include "PRED_reverse_2.h"
#include "../../../machine/SxxMachine/SymbolTerm.h"
#include "../../../machine/SxxMachine/Operation.h"
#include "../../../machine/SxxMachine/Term.h"
#include "../../../machine/SxxMachine/Prolog.h"
#include "../../../machine/SxxMachine/ListTerm.h"
#include "../../../machine/SxxMachine/TermData.h"

namespace SxxMachine
{
	using ListTerm = SxxMachine::ListTerm;
	using Operation = SxxMachine::Operation;
	using Predicate = SxxMachine::Predicate;
	using Prolog = SxxMachine::Prolog;
	using SymbolTerm = SxxMachine::SymbolTerm;
	using Term = SxxMachine::Term;
	using TermData = SxxMachine::TermData;
SxxMachine::SymbolTerm *const PRED_reverse_2::s1 = SxxMachine::SymbolTerm::intern("[]");

	PRED_reverse_2::PRED_reverse_2(Term *a1, Term *a2, Operation cont)
	{
		this->LARG[0] = a1;
		this->LARG[1] = a2;
		this->cont = cont;
	}

	void PRED_reverse_2::toString(StringBuilder *sb)
	{
		sb->append("reverse(");
//JAVA TO C++ CONVERTER TODO TASK: There is no native C++ equivalent to 'toString':
		LARG[0]->toString(sb);
		sb->append(", ");
//JAVA TO C++ CONVERTER TODO TASK: There is no native C++ equivalent to 'toString':
		LARG[1]->toString(sb);
		sb->append(")");
	}

	Operation PRED_reverse_2::exec(Prolog *engine)
	{
		engine->setB0();
		Term *a1 = LARG[0]->dref();
		Term *a2 = LARG[1]->dref();

		if (dynamic_cast<ListTerm*>(a1) != nullptr || a1->isNil())
		{
			Term *r = Prolog::Nil;
			while (dynamic_cast<ListTerm*>(a1) != nullptr)
			{
				r = TermData::CONS(a1->asListTerm()->car(), r);
				a1 = a1->asListTerm()->cdr();
			}
			if (a1->isNil() && a2->unify(r, engine->trail))
			{
				return cont;
			}
		}
		return engine->fail();
	}
}
